### PR TITLE
Add QuorumCertificate method to Snapshot

### PR DESCRIPTION
**Pull request by jordanschalm**
This PR adds `QuorumCertificate` method to `protocol.Snapshot`. This isn't strictly necessary for serializable snapshots, however we needed at least the signature from the QC in order to satisfy the `Seed` method, and we wanted to make this change anyway (see https://github.com/dapperlabs/flow-go/issues/4892).

This PR does not handle the case where the quorum certificate is requested for the root snapshot, but should handle all other cases. Storing the root QC will be handled as part of the `Bootstrap` changes in https://github.com/dapperlabs/flow-go/issues/4890.

### CODE REVIEW

File: state/protocol/badger/snapshot.go [hunk 1]
```diff
@@ -41,6 +41,88 @@ func (s *Snapshot) Head() (*flow.Header, error) {
...
+		// TODO store root QC and return here
+		return nil, fmt.Errorf("root qc not stored")
+	}
+
+	// CASE 2: for any other block, generate the root QC from a valid child
+	child, err := s.validChild()
+	if err != nil {
+		return nil, fmt.Errorf("could not get child: %w", err)
+	}
+
```
Comment (zhangchiqing) [inline state/protocol/badger/snapshot.go:71]: Better to add a sanity check that the `header.ID() == child.ParentID`

**Review by zhangchiqing — APPROVED** [review]
Looks good 👍

**Review by AlexHentschel — APPROVED** [review]
Looks great. 

Something not directly related to your PR, which I noticed:
* `Snapshot.Seed(...)` and `Snapshot.QuorumCertificate()` directly work with the data base. There is no caching layer in between. This concerns me, as we had previously significant issues with directly accessing the data base:
    - it cost us two days to benchmark until Max Wolter narrowed the problem down to the protocol state taking way too long when it was always directly accessing the DB
    - As a result, Max introduced caching layers for all storage access in the protocol state. 
* I feel it would be a relatively moderate time investment now, to also cache a block's valid child (e.g. in the mutator as part of the MakeValid call). In contrast, if we don't do it now and it becomes a performance problem, it will likely cost us significantly more time to only  identify that finding a valid child is taking too long through benchmarking.  

Totally happy, if we do this in a separate PR. Just wanted to flag it.

File: state/protocol/badger/snapshot.go [hunk 2]
```diff
@@ -41,6 +41,88 @@ func (s *Snapshot) Head() (*flow.Header, error) {
...
+	var childIDs []flow.Identifier
+	err := s.state.db.View(procedure.LookupBlockChildren(s.blockID, &childIDs))
+	if err != nil {
+		return nil, fmt.Errorf("could not look up children: %w", err)
+	}
+
+	// check we have at least one child
+	if len(childIDs) == 0 {
+		return nil, state.NewNoValidChildBlockError("block doesn't have children yet")
+	}
```
Comment (AlexHentschel) [inline state/protocol/badger/snapshot.go:97]: Is this check necessary here? I feel that the extra cost of looping over an empty slice is negligible

File: state/protocol/badger/snapshot.go (hunk 1 above)
Comment (jordanschalm) [reply state/protocol/badger/snapshot.go:71]: 👍 Added in https://github.com/onflow/flow-go/pull/201/commits/7f383638d3cf618bf861e01a3e649c6f3b7b64a3

File: state/protocol/snapshot.go [hunk 3]
```diff
@@ -19,6 +19,10 @@ type Snapshot interface {
 	// we should build the next block in the context of the selected state.
 	Head() (*flow.Header, error)
 
+	// QuorumCertificate returns a valid quorum certificate for the header at
+	// this snapshot, if one exists.
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/snapshot.go:23]: Done in https://github.com/onflow/flow-go/pull/201/commits/8f5992ea9cfc8c97ec471b5fe59c39eafcf7fa26

File: state/protocol/badger/snapshot.go [hunk 4]
```diff
@@ -41,6 +41,88 @@ func (s *Snapshot) Head() (*flow.Header, error) {
...
+	root, err := s.state.Params().Root()
+	if err != nil {
+		return nil, fmt.Errorf("could not get root: %w", err)
+	}
+
+	head, err := s.Head()
+	if err != nil {
+		return nil, fmt.Errorf("could not get head: %w", err)
+	}
+	if head.Height == root.Height {
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/badger/snapshot.go]: Done in https://github.com/onflow/flow-go/pull/201/commits/f3ca5f1461153d55905e40cf6fe418eda7e800b1. Changed this for `Epochs().Previous()` and `Identities()` as well where we were previously comparing using `Height`.

File: state/protocol/badger/snapshot.go [hunk 5]
```diff
@@ -41,6 +41,88 @@ func (s *Snapshot) Head() (*flow.Header, error) {
...
+		BlockID:   child.ParentID,
+		SignerIDs: child.ParentVoterIDs,
+		SigData:   child.ParentVoterSig,
+	}
+
+	return qc, nil
+}
+
+// validChild returns a child of the snapshot head that has been validated
+// by HotStuff, if such a child exists. Otherwise returns an error. Any valid
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/badger/snapshot.go]: https://github.com/onflow/flow-go/pull/201/commits/8ebbb8c7b04da86fb537f58eb605b53321d73377

File: state/protocol/badger/snapshot_test.go [hunk 6]
```diff
@@ -130,19 +130,27 @@ func TestClusters(t *testing.T) {
...
+			err = state.Mutate().Extend(&block1)
+			require.Nil(t, err)
 
-			_, err = state.Final().(*bprotocol.Snapshot).Seed(1, 2, 3, 4)
-			t.Log(err)
+			_, err = state.AtBlockID(block1.ID()).QuorumCertificate()
+			assert.Error(t, err)
+
+			_, err = state.AtBlockID(block1.ID()).Seed(1, 2, 3, 4)
 			assert.Error(t, err)
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/badger/snapshot_test.go:156]: https://github.com/onflow/flow-go/pull/201/commits/a490ba83ff4d77be97b2f51c1db92846a1c48ad2

File: state/protocol/badger/snapshot_test.go [hunk 7]
```diff
@@ -130,19 +130,27 @@ func TestClusters(t *testing.T) {
...
+			require.Nil(t, err)
+			block1 := unittest.BlockWithParentFixture(root.Header)
+			block1.SetPayload(flow.EmptyPayload())
+			err = state.Mutate().Extend(&block1)
+			require.Nil(t, err)
 
-			_, err = state.Final().(*bprotocol.Snapshot).Seed(1, 2, 3, 4)
-			t.Log(err)
+			_, err = state.AtBlockID(block1.ID()).QuorumCertificate()
+			assert.Error(t, err)
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/badger/snapshot_test.go:152]: https://github.com/onflow/flow-go/pull/201/commits/a490ba83ff4d77be97b2f51c1db92846a1c48ad2

File: state/protocol/badger/snapshot.go (hunk 2 above)
Comment (jordanschalm) [reply state/protocol/badger/snapshot.go]: 👍  Good point https://github.com/onflow/flow-go/pull/201/commits/ff6f97cf8f1e40bf9a7009541739908c6cb4b0ff

**Comment by jordanschalm** [comment]
> Snapshot.Seed(...) and Snapshot.QuorumCertificate() directly work with the data base. There is no caching layer in between.

It's true. On the other, retrieving the latest finalized or sealed block, an operation we do very often currently, also goes directly to Badger (twice actually!):
https://github.com/onflow/flow-go/blob/9d69ff9da85ca942a519b34887f47c7d2cdad788/state/protocol/badger/state.go#L91-L99

Badger does have a built-in cache, but the cache contains the serialized data. My understanding is that for some larger models (and for the special case of BLS keys, where deserialization involved expensive cryptographic checks), deserialization was expensive enough to warrant another caching layer. I assumed for small DB values like heights and indexes the deserialization cost was small enough we got most of the benefits from Badger's built-in cache. 

However, it turns out that Badger disabled caching by default in [`v2.0.3`](https://github.com/dgraph-io/badger/pull/1257), which we switched to back in June ([here](https://github.com/onflow/flow-go/commit/c072de0f84a43f272d15b60ce8a2e4876b168595#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R11) as far as I can tell). Apparently having the cache turned on is slower than having it turned it for the default configuration 🤔.

For now I made an issue for adding the cache as you suggested. I also think we should add tracing to a few of these paths that hit the database to get a sense for the current cost. https://github.com/dapperlabs/flow-go/issues/5115

**Pull request merged**